src/range.cr

```diff
@@ -392,12 +392,78 @@ struct Range(B, E)
       raise ArgumentError.new("Can't sample an open range")
     end
 
-    return super unless n == 1
+    if n < 0
+      raise ArgumentError.new "Can't sample negative number of elements"
+    end
 
-    if empty?
+    case n
+    when 0
       [] of B
-    else
+    when 1
       [sample(random)]
+    else
+      # For a range of integers we can do much better
+      {% if B < Int && E < Int %}
+        min = self.begin
+        max = self.end
+
+        if (exclusive? && max <= min) || (!exclusive? && max < min)
+          raise ArgumentError.new "Invalid range for rand: #{self}"
+        end
+
+        max -= 1 if self.exclusive?
+
+        available = max - min + 1
+        possible = Math.min(n, available)
+
+        # If we must return all values in the range...
+        if possible == available
+          result = Array(B).new(possible)
+          each { |value| result << value }
+          result.shuffle!
+          result
+        elsif n <= 16
+          # For a small requested amount doing a linear lookup is faster
+          result = Array(B).new(possible)
+          add_n_samples(result, possible, random)
+          result
+        else
+          # Otherwise using a Set is faster
+          result = Set(B).new(possible)
+          add_n_samples(result, possible, random)
+          result.to_a
+        end
+      {% elsif B < Float && E < Float %}
+        min = self.begin
+        max = self.end
+
+        if (exclusive? && max <= min) || (!exclusive? && max < min)
```

Comment:
Isn't this equvivalent to `max < min || (exclusive? && max == min)`?

Reply:
`x...x` usually denotes an empty range, so maybe that should be valid for `#sample(n)` and return an empty array? It's an edge case.

Reply:
In my mind `x...x` doesn't include `x` so you can't pick it. For example `(0...0).to_a` is an empty array.

Reply:
`(0...0).sample(10)` should be an empty array, too.

Reply:
Oh, that's what you mean... mmm... but if you do `(0...0).sample` you get an error. That's why I'm not sure it should work in one case and not in the other.

Reply:
You could see it as this: `1...2` results in an array with one element, `1...1` is just "one less" i.e. an empty array. I'm not totally convinced either, though.

Reply:
The most optimal approach (and the most direct conversion of this snippet) is
 `(exclusive? ? max <= min : max < min)`

Reply:
It's true. I don't know what else can we use to take a decision on this.